### ISSUE TITLE
[selectors] Let :matches() have better error-recovery behavior than normal Selectors

### ISSUE THREAD

**Issue by tabatkins** [w3c/csswg-drafts#3264]
This isn't captured in the current spec, but I remember earlier speculation that we could use `:matches()` as a way to get around the bad Selectors behavior of "a syntax error in one complex selector invalidates the whole sequence" that we're stuck with.  Do we still think this is a worthwhile idea to pursue?

Spec-wise, what this would mean is defining the official syntax as `:matches( <any-value> )`, then split the result on top-level comma tokens, then attempt to parse each item as a `<complex-selector>`, and just ignore any invalid ones. (If all of them are invalid, the selector matches nothing.)

Then, if you're concerned about using a newer feature, you can just write your selectors like:

```
:matches( .foo || .bar, td.foo:first-child ) {
  /* styles for the bar cell */
```

An unfortunately, but relatively minor, tax for getting better error-recovery behavior. (If we *do* end up renaming it to `:is()`, it's even more minor.)

**Comment by CyberAP** [comment]
Shouldn't this belong to `@supports` rule?
Just to imagine how bloated stylesheet would get if you have to use new selectors extensively. While with `@supports` you can only test it once and that's it.
Yes it doesn't cover selector check right now, but with [Custom Selectors from CSS Extensions](https://drafts.csswg.org/css-extensions/#custom-selectors) this should be possible.

**Comment by SelenIT** [comment]
@CyberAP, the point is to avoid having to write the same rules twice (for new and all browsers).

For example, consider a menu that is by default `transform`ed into a hamburger icon (or something), but should "uncollapse" if either hovered or has a link inside it focused via keyboard. For modern browsers, I can write
```css
.menu:hover, .menu:focus-within { transform: none; }
```
...but in browsers that don't support `:focus-within` this would not apply at all, and I would have to write this rule twice, for each selector separately, in order to make at least the `:hover` part work there.

However, with the new proposal, the following hypothetical example
```css
:is(.menu:hover, .menu:focus-within, :current(.menu)) { transform: none; }
```
would be able to work in browsers that already support `:is()` and these two pseudo-classes, but don't support time-dimensional pseudo-classes yet.

How can this be solved with `@supports` (even if extended to test for selectors)? For me, this looks like the opposite of what `@supports` does (combining the code for old and new browsers into one rule, rather than separating it into different rules). Would be happy to be proven wrong!

**Comment by CyberAP** [comment]
@SelenIT, thanks for a great explanation! That does make sense and actually is a really great feature I would like to have.

**Comment by ExE-Boss** [comment]
You mean like https://github.com/w3c/csswg-drafts/issues/3082, except without the vendor prefix exemption?

**Comment by tabatkins** [comment]
Yes, and safely; I'm pretty sure we can't do #3082.

**Comment by fantasai** [comment]
emilio pointed out that we need to be careful about how we handle `:not()` here. It's not obvious it would work as intended if we just ignore unknown arguments, and also a single argument to `:not()` likely has Web-compat implications since it's already implemented.

A related question is `:nth-child()`; simply ignoring one of its arguments could throw off the count. There likely wouldn't be a Web-compat problem with it, though.

**Comment by ExE-Boss** [comment]
The thing about `:not()` is that only the single&#x2011;argument form is supported in major browsers, so we could probably  apply it to the multi&#x2011;argument form.

**Comment by ewilligers** [comment]
I agree that we could minimize the Web-compat risk by saying that error recovery only applies if two or more selectors are supplied.

`:nth-last-child(:nonsense)` is invalid
`:nth-child(:nonsense, :enabled)` is `:nth-child(:enabled)`
`:not(:nonsense, :gibberish)` is `*`
`:is(:nonsense, :gibberish)` is `:not(*)`.

This would have the following benefits:

* `:not(.)` and `:nth-child(.)` and `:nth-last-child(.)` with a single selector have the same semantics as in Selectors 3.
* We have error recovery for `:is(...)` and `:where(...)` and `:has(...)`
* `:not(...)` and `:is(:not(...))` and `:not(:is(...))` are always equivalent, same for `:nth-child` and `:nth-last-child`

As a minor variation, preserving the same benefits, we could say that error recovery only applies if at least one of the selectors is valid. Then

`:not(:nonsense, :gibberish)` is invalid.

This might be easier to teach, and avoid developers' confusion when they see an introduced `:not(*)` in a Developer Tool that hasn't preserved raw text.

**Comment by ExE-Boss** [comment]
> As a minor variation, preserving the same benefits, we could say that error recovery only applies if at least one of the selectors is valid.
>
>Then `:not(:nonsense, :gibberish)` is invalid.

That should probably only apply to `:not(…)`, as `:commonly-supported, :is(:rarely-supported)` can be used to avoid invalidating the entire selector list if the `:rarely-supported` pseudo‑class isn’t supported by the current browser but `:is(…)` and `:commonly-supported` are supported by the current browser.

Also, it might be a good idea to still parse `:not(:nonsense, :gibberish)` as valid, but matching nothing.

**Comment by ewilligers** [comment]
For `:not()`, I [added](https://chromium-review.googlesource.com/c/chromium/src/+/1354751) Blink use counters.

[kCSSSelectorNotWithValidList](https://www.chromestatus.com/metrics/feature/timeline/popularity/2645)
6% of pages have `:not()` with 2+ selectors, all valid, or a valid non-simple selector,  e.g. `:not(*, *)` or `:not(* + *)`

[kCSSSelectorNotWithInvalidList](https://www.chromestatus.com/metrics/feature/timeline/popularity/2646)
4.5% of pages have `:not()` with no valid selectors, e.g. `:not(:nonsense)`

[kCSSSelectorNotWithPartiallyValidList](https://www.chromestatus.com/metrics/feature/timeline/popularity/2647)
0.09% of pages have `:not()` containing both valid and invalid selectors, e.g. `:not(*, :nonsense)`

**Comment by ewilligers** [comment]
We haven't discussed what dev tools, etc., should show when the author provided `:is(:nonsense)`. Do people prefer `:is()` or `:is(:not(*))`? This might not be a spec issue.

**Comment by ExE-Boss** [comment]
I&nbsp;think <code>:is(~~:nonsense~~)</code> might be&nbsp;best.

**Comment by ExE-Boss** [comment]
What should happen if someone were to write:
```css
:is(h1, h3), :is(h2:maybe-unsupported) { font-family: sans-serif }
```

I believe that it should be treated as if the following was written in browsers that don't support `:maybe‑unsupported`, but support `:is(…)`:
```css
:is(h1, h3) { font-family: sans-serif }
```

**Comment by pygy** [comment]
I'm all for this, but it will interact surprisingly with selector nesting, since `a, b { & c, &d {}}` becomes `:is(a, b) c, :is(a, b), d{}`.

It makes nesting forgetful except at the last level... So `c` or `d` resulting in a parsing error will cause the rejection of the last nesting level.

**Comment by danburzo** [comment]
Drive-by comment: seems that in Firefox 78+, and the upcoming Safari 14 (currently in the Technology Preview stage), the `:is()` and `:where()` pseudo-classes become invalid when an invalid selector is included in the list. For example, `document.querySelectorAll('div:is(div, :bogus)')` throws in both.

**Comment by frivoal** [comment]
Agenda+ to look into the problem raised by https://github.com/w3c/csswg-drafts/issues/3264#issuecomment-675127431, and to consider whether compat is starting to paint us into a corner here.

**Comment by emilio** [comment]
I'm happy to implement this in Gecko if people define in the spec how to deal with the empty-selector behavior. Like, should `:is(:nonsense)` serialize just as `:is()`? Should that be special-cased to never match?

Also, it may be a bit confusing if people write stuff like `:is(::before, ::after)`, and it gets turned to `:is()` instead of being rejected as invalid...

**Comment by emilio** [comment]
Then, of course `:is()` should be parseable too...

**Comment by ewilligers** [comment]
> should `:is(:nonsense)` serialize just as `:is()`? Should that be special-cased to never match?

> Then, of course `:is()` should be parseable too...

Yes, I was going to serialize `:is(:nonsense)` as `:is()` in dev tools, and treat `:is()` as a valid selector that never matches.

Is serialization of selectors web exposed, i.e. testable in WPTs?

**Comment by Loirooriol** [comment]
@ewilligers Sure, include `/css/support/parsing-testcommon.js` and then

```js
test_valid_selector(":is(:nonsense)", ":is()");
```

**Comment by emilio** [comment]
https://bugzilla.mozilla.org/show_bug.cgi?id=1664718 is the Gecko bug for this (sorry for the lag, busy month!). I plan to enable it on Nightly and beta for now to collect potential breakage, but will enable after a couple releases everywhere if there's no issue.

**Comment by tabatkins** [comment]
Just verifying as I write the spec - in <https://github.com/w3c/csswg-drafts/issues/3264#issuecomment-441148706> @ewilligers suggested that we handle "single selector, which is invalid" differently from the rest of the cases, since that has legacy implications, but it looks like we're no longer doing that?

That is, `:is(:unknown)` is valid but matches nothing, just like `:is()` and `:is(:unknown1, :unknown2)`?

**Comment by emilio** [comment]
Yes, that's what I implemented, fwiw. I don't see why "single invalid selector" should be different.

**Comment by tabatkins** [comment]
Cool, then please make sure the spec edits linked above are satisfactory; they apply to :is(), :where(), and :has(), but not :not(), per resolutions earlier in the thread.

**Comment by emilio** [comment]
They look good. I'd probably note that there's an interesting side-effect of this behavior which I'm not sure it's great, which is that the specificity of `:is(:unknown, div)`, will change when a browser starts supporting `:unknown`. I don't see a great way to avoid that though.

**Comment by tabatkins** [comment]
Yeah, unless I distinguish between "parsable, just using unknown selectors" and "lol wtf is this", I can't handle that.

And even then, a "parsable but unknown" selector might have complex specificity behavior, so we actually can't even predict that. So it's just something we'll have to accept.
